### DISCUSSION TITLE
ENSO use cases?

### DISCUSSION THREAD

**Discussion by benjamin-cash** [dtcenter/METplus#2222] [Use Cases] [answered]
Are there any use cases for calculating ENSO indices, verifying ENSO forecasts, calculating teleconnections, etc?

**Comment by CPKalb** [comment]
Hi Benjamin:

[This](https://metplus.readthedocs.io/en/latest/generated/model_applications/s2s_mjo/UserScript_obsCFSR_obsOnly_MJO_ENSO.html#sphx-glr-generated-model-applications-s2s-mjo-userscript-obscfsr-obsonly-mjo-enso-py) is an ENSO use case that computes the MaKE and MaKI indices.  I see that it doesn't have ENSO in the title, so we should probably add that for more clarity.

Currently, that is the only ENSO use case we have.  I have done some work with ENSO in the past (before METplus).  Are you looking for a use case that does something specific?

Christina

**Reply by benjamin-cash** [reply]
Hi @CPKalb - I was looking for something that would do a basic verification of an ENSO plume. I was also interested in something that would compute one point correlations between, say, nino34 and 200mb height, for model and observations and compare the patterns (ie some kind of assessment of the ENSO teleconnections). Poking around it doesn't seem like anything any of the use cases currently do this?

**Reply by TaraJensen** [reply]
Ben, you are correct.  We don't have those basic verification examples because we do not have a true S2S expert our team.  Would you be willing to have a telecon to discuss what you mean by "ENSO plume" and "one point correlations between nino34 and 200mb"?  Or point us to a paper or module in MDTF, that does what you're looking for? I'm guessing we might be able to construct use cases that address these basic needs, but we need more info.

**Reply by benjamin-cash** [reply]
Hi Tara, sure thing. :)

In the meantime, just to expand a little on what I meant above, by 'ENSO plume' I meant something like the following. Say I have an ensemble of N model runs, where each run is M months long. I want to calculate the NINO34 index for each ensemble member for each of my start dates and then compare that to observations through one of the ensemble/probabilistic metrics. I think that this is potentially something that is similar to existing use cases and maybe just requires EnsembleStat, and either GridStat to calculate NINO34 (can you specify a lat-lon region to perform the average over?) or a UserScript?

Following up, it looks like some of the functionality for calculating the basic ENSO properties is probably available from here: https://mdtf-diagnostics.readthedocs.io/en/latest/sphinx_pods/ENSO_MSE.html

**Reply by CPKalb** [reply]
Hi Ben:

I'm definitely rusty on ENSO 3.4.  But I believe it requires computation of anomalies.  This has been done using pcp_combine as part of the blocking use case, although it is in one of the steps that are commented out.

You are correct.  It would probably be Ensemble-Stat that is needed to compute ensemble statistics.  For the lat-lon, do you need to perform verification over all points within the lat-lon region or create an average over the whole region?  The former is easy in MET, but I don't believe we have a tool that does the latter.

Christina

**Reply by benjamin-cash** [reply]
Hi @CPKalb - NINO34 (or most of the other indices) are calculated as the area average anomaly over a given box in the tropical Pacific. It seems like maybe this could be done using a mask and GridStat?

**Reply by CPKalb** [reply]
Hi Ben:

You can get means and anomalies from Grid-Stat (output from SL1L2, or SAL1L2 lines), but they are one number across the entire domain.  It sounds like this is what you want.  However, the issue would be computing statistics on that output.  Ensemble-Stat requires gridded model data to run, and would not work on a single number.  Stat-Analysis can be used to compute statistics on single numbers as matched pairs, but ensemble statistics are not possible from the SL1L2 and SAL1L2 lines, or even from the matched pair (MPR) line type.  To get ensemble statistics from Stat Analysis on this single data point, the data would somehow need to be put in the format of the [ORANK line output](https://met.readthedocs.io/en/latest/Users_Guide/ensemble-stat.html#id7).  This [table ](https://met.readthedocs.io/en/latest/Users_Guide/stat-analysis.html#job-aggregate-stat) gives information about how Stat-Analysis can produce aggregate statistics from different line types.  The only ensemble entry on the table is for the ORANK line.

Did that make any sense?  I'm not sure I'm explaining this well.

Christina

**Reply by benjamin-cash** [reply]
Hi Christina - How does Ensemble-Stat determine if a file is gridded? For example, if my timeseries of NINO34 values was formatted to include a lat-lon value, even if it was just a single pair, would the tool accept it?

**Reply by CPKalb** [reply]
HI Ben:

Ensemble-Stat is set up to read a gridded file in either netCDF (CF compliant, MET format, or from p_interp), grib1/grib2, or from python embedding.  In all of these cases, Ensemble-Stat would expect information about the grid (see this [information](https://met.readthedocs.io/en/latest/Users_Guide/appendixB.html) on supported grids).  There are specific variables defining the grid that are required, and they vary depending on what projection is used.  My suspicion is that giving Ensemble-Stat one data value with a lat/lon associated would not work, though I have not actually tested this.  I have, in the past, tried to provide MET with a grid projection that is not supported and it fails with an error.

@JohnHalleyGotway is probably the best person to provide more information on how gridded data is actually read in to MET.  However, he is on vacation at the moment.  I can check here to see if I can get more information.

Christina

**Reply by benjamin-cash** [reply]
Thanks! I'll keep an eye on this space. :)

**Reply by benjamin-cash** [reply]
Is John back in the office this week or still out on vacation?

**Reply by JohnHalleyGotway** [reply, marked as answer]
Sorry for the long delay on this. I didn't have it on my radar.

So the question is how trivial of a "grid" can MET actually handle? For example, could it handle a "grid" consisting of a single point? I don't know the answer to that off-hand. I'd say maybe, but perhaps there's a better way of handling this situtation?

This reminds me a lot of the case where we already have forecast and observation data matched up at station locations. Point-Stat compares *gridded* forecasts against to *point* observations. But if you already have a point forecast matched to a point observation, what should you do? In the past, we've used this data to populate the MPR .stat line type, as if it were output generated from Point-Stat. We reformat that paired data into MPR lines and run the Stat-Analysis tool to compute statistics for those pairs. @CPKalb, I recall you taking this approach for the Xcel Energy project in the past.

The exact same approach can be used with ensemble data, but instead of populating the MPR line type, you populate the [ORANK line type](https://met.readthedocs.io/en/latest/Users_Guide/ensemble-stat.html?highlight=orank#id7). Once the point ensemble forecast and observations are reformatted into ORANK lines for one or more locations and times, you pass them as input to Stat-Analysis and then recompute whatever other ensemble output you want.

Does that sound like what you're looking for? Or should I figure out if MET can handle a 1x1 size grid?

**Reply by benjamin-cash** [reply]
Hi @JohnHalleyGotway - Now I get to apologize for taking a while to get back to this. :)

Maybe it makes more sense to take a step back and describe what I am trying to do a little better. I have an ensemble of timeseries from a set of model runs. I want to use some of the METplus ensemble/probabilistic verification tools to assess my model against observations. Maybe I could use python embedding to calculate the index in the ORANK format and then feed it into Stat-Analysis? Ideally I would have something that would work on netcdf data (or xarray via embedding) so that I didn't need to transform it into a different data type.

**Reply by CPKalb** [reply]
Hi John and Ben:

Yes, I have taken this type of approach with the Kuwait energy project in the past.  In this case, I had model data that was a single point and wanted to compare it to observations that were also a single point.  MET generally requires comparing a gridded model to point data (point_stat, ensemble_stat) or a gridded model to gridded observations (grid_stat, ensemble_stat). To get around this, I used an external python script (not connected to MET) to put data in the MPR line format and fed that into Stat-Analysis.  So, the only MET program I ran here was Stat-Analysis.

Since Stat-Analysis can now handle [python embedding for the MPR line format](https://met.readthedocs.io/en/latest/Users_Guide/appendixF.html#python-embedding-for-mpr-data), I didn't write any output mpr files, but rather just used python embedding to get these into Stat-Analysis.  

However, I'm not sure if Stat-Analysis has python embedding support for the ORANK line type.   @JohnHalleyGotway, can Stat-Analysis read the ORANK line type from python embedding or would a file need to be written with the ORANK line type?

Tina

**Reply by JohnHalleyGotway** [reply]
@CPKalb I realize that this discussion has remained unanswered for a few months now. I believe that we've followed up on these detail in meetings and perhaps on Slack. But I wanted to clarify the state of things in this issue.

Basically, no, as of right now, METplus is not setup well to serve any .stat line type to Stat-Analysis via python. Currently, only the use of MPR data is well supported. However, the dtcenter/MET#2539 issue requests exactly that functionality. I'll note that no changes are actually needed in Stat-Analysis itself. Instead, changes are needed in to the python wrapper code in MET.

I recommend that we mark this discussion as being answered and move any future discussion about python embedding of .stat data in the Stat-Analysis tool to comments on that dtcenter/MET#2539 issue. If that's acceptable, can one of you please select a good answer from this history of correspondence?

**Comment by TaraJensen** [comment]
He's out until the 10th.
